### PR TITLE
add pinot upsert features to pinot common

### PR DESCRIPTION
**Pull request by jamesyfshao**
This diff will create necessary pinot schema/constant value changes to support pinot upsert features.

The following will be changed:

1. add IngestionModeConfig field to schema and table definition:
This field will contains several sub-fields:
   a. ingestionMode: 
ingestion mode could be either upsert/append.
  b. primaryKeys:
a list of string that indicate which columns would be use as primary key of the current upsert table. However we are going to enforce primary key to be one only only one column. In the future we can explore the support for multiple primary keys
  c. offsetKey:
a string that indicate which column would be use as offset of the current pinot upsert table.

design doc for pinot upsert feature:

https://cwiki.apache.org/confluence/display/PINOT/%5BProposal%5D+Pinot+upsert+design+doc
this change focus on the changes in pinot common repo

### CODE REVIEW

File: pinot-common/src/main/java/org/apache/pinot/common/config/TableConfig.java [hunk 1]
```diff
@@ -75,6 +79,9 @@
   private Map<InstancePartitionsType, InstanceAssignmentConfig> _instanceAssignmentConfigMap;
   private List<FieldConfig> _fieldConfigList;
 
+  @JsonPropertyDescription(value = "The update semantic of the table, either append or upsert, default as append")
+  private UpdateSemantic _updateSemantic;
```
Comment (kishoreg) [inline pinot-common/src/main/java/org/apache/pinot/common/config/TableConfig.java]: rename to IngestMode or just have a boolean enableUpsert?

File: pinot-common/src/main/java/org/apache/pinot/common/config/TableNameBuilder.java [hunk 2]
```diff
@@ -132,4 +132,19 @@ public static boolean isOfflineTableResource(String resourceName) {
...
     return REALTIME.tableHasTypeSuffix(resourceName);
   }
+
+  /**
+   * ensure that table name ends with type info, if no, create one with the given type
+   * @param tableName the name of the table
+   * @param type the type of the table for it to fill in if the type info is missing
+   * @return the table type name with the type info
+   */
+  public static String ensureTableNameWithType(String tableName, TableType type) {
```
Comment (kishoreg) [inline pinot-common/src/main/java/org/apache/pinot/common/config/TableNameBuilder.java]: where is this used?

File: pinot-common/src/main/java/org/apache/pinot/common/metrics/BrokerGauge.java [hunk 3]
```diff
@@ -26,7 +26,8 @@
  *
  */
 public enum BrokerGauge implements AbstractMetrics.Gauge {
-  QUERY_QUOTA_CAPACITY_UTILIZATION_RATE("tables", false), NETTY_CONNECTION_CONNECT_TIME_MS("nettyConnection", true);
+  QUERY_QUOTA_CAPACITY_UTILIZATION_RATE("tables", false), NETTY_CONNECTION_CONNECT_TIME_MS("nettyConnection", true),
+  TABLE_MIN_LOW_WATER_MARK("tables", false);
```
Comment (kishoreg) [inline pinot-common/src/main/java/org/apache/pinot/common/metrics/BrokerGauge.java]: what does this represent?

File: pinot-spi/src/main/java/org/apache/pinot/spi/data/Schema.java [hunk 4]
```diff
@@ -59,13 +65,28 @@
   private static final Logger LOGGER = LoggerFactory.getLogger(Schema.class);
 
   private String _schemaName;
+
+  // upsert related config
+  // primary key refers to the column name that is the primary key of this upsert table
+  private String _primaryKey;
```
Comment (kishoreg) [inline pinot-spi/src/main/java/org/apache/pinot/spi/data/Schema.java]: what if the primary key is made up of two columns?

Comment (kishoreg) [reply pinot-spi/src/main/java/org/apache/pinot/spi/data/Schema.java]: or multiple

File: pinot-spi/src/main/java/org/apache/pinot/spi/data/Schema.java [hunk 5]
```diff
@@ -59,13 +65,28 @@
...
+  // offset key refers to the column name that we are going to store the offset value to
+  private String _offsetKey;
```
Comment (kishoreg) [inline pinot-spi/src/main/java/org/apache/pinot/spi/data/Schema.java]: do we need all these things in Schema? What if we create a UpsertConfig in TableConfig to provide all the necessary information.

File: pinot-spi/src/main/java/org/apache/pinot/spi/data/Schema.java [hunk 6]
```diff
@@ -631,4 +751,50 @@ public int hashCode() {
...
   }
+
+  public boolean isVirtualColumn(String columnName) {
+    return columnName.startsWith("$") || (getFieldSpecFor(columnName).getVirtualColumnProvider() != null
+        && !getFieldSpecFor(columnName).getVirtualColumnProvider().isEmpty());
+  }
+
+
+  @JsonIgnore
+  public static byte[] getByteArrayFromField(Object value, DimensionFieldSpec fieldSpec) {
```
Comment (kishoreg) [inline pinot-spi/src/main/java/org/apache/pinot/spi/data/Schema.java]: why is this needed? it will be great if we can avoid static methods in config objects. We can probably move this SchemaUtil

File: pom.xml [hunk 7]
```diff
@@ -132,6 +132,9 @@
     <log4j.version>2.11.2</log4j.version>
     <netty.version>4.1.42.Final</netty.version>
 
+    <!-- dependency for upsert component -->
```
Comment (kishoreg) [inline pom.xml]: Do we need this at compile time?

File: pinot-common/src/main/java/org/apache/pinot/common/config/TableConfig.java [hunk 8]
```diff
@@ -30,17 +30,21 @@
 import java.util.List;
 import java.util.Map;
 import javax.annotation.Nullable;
+import javax.validation.constraints.Null;
```
Comment (Jackie-Jiang) [inline pinot-common/src/main/java/org/apache/pinot/common/config/TableConfig.java]: Wrong import

File: pinot-common/src/main/java/org/apache/pinot/common/config/TableConfig.java [hunk 9]
```diff
@@ -354,6 +387,7 @@ public ZNRecord toZNRecord()
     }
 
     ZNRecord znRecord = new ZNRecord(_tableName);
+    simpleFields.put(UPDATE_SEMANTIC_CONFIG_KEY, _updateSemantic.toString());
```
Comment (Jackie-Jiang) [inline pinot-common/src/main/java/org/apache/pinot/common/config/TableConfig.java]: This can cause NPE

File: pinot-common/src/main/java/org/apache/pinot/common/config/TableConfig.java (hunk 1 above)
Comment (Jackie-Jiang) [reply pinot-common/src/main/java/org/apache/pinot/common/config/TableConfig.java]: Is this for real-time? We already have push type of either APPEND or REFRESH for offline; aggregateMetrics for real-time. What does UPSERT stand for?

File: pinot-common/src/main/java/org/apache/pinot/common/config/TableNameBuilder.java (hunk 2 above)
Comment (Jackie-Jiang) [reply pinot-common/src/main/java/org/apache/pinot/common/config/TableNameBuilder.java]: Please use `TableNameBuilder.forType(type).tableNameWithType(tableName)`

File: pinot-common/src/main/java/org/apache/pinot/common/metrics/AbstractMetrics.java [hunk 10]
```diff
@@ -21,16 +21,17 @@
 import com.google.common.annotations.VisibleForTesting;
 import com.yammer.metrics.core.MetricName;
 import com.yammer.metrics.core.MetricsRegistry;
+import org.apache.pinot.common.Utils;
```
Comment (Jackie-Jiang) [inline pinot-common/src/main/java/org/apache/pinot/common/metrics/AbstractMetrics.java]: Wrong order, please revert

File: pinot-common/src/main/java/org/apache/pinot/common/config/TableConfig.java (hunk 1 above)
Comment (jamesyfshao) [reply pinot-common/src/main/java/org/apache/pinot/common/config/TableConfig.java]: @kishoreg I am thinking if using enum here will be more flexible in the case of future features changes or whatnot. Boolean seems to be only limiting two options and it might limit what other changes we might want to do with pinot. Please let me what you think. ingestionMode definitely sounds more explicit than what I am using here and I will definite consider about this.

@Jackie-Jiang yes. So basically it is previous old design about we want to support update in pinot realtime ingestion by adding override by primary key (eg, a second message in Kafka with the same primary key represents the update of existing key instead of two records with the same primary key). Upsert stands for this new ingestion semantic (insert if no duplicated primary key, update if duplicate primary key). This feature will be applying to realtime table only

File: pinot-common/src/main/java/org/apache/pinot/common/metrics/BrokerGauge.java (hunk 3 above)
Comment (jamesyfshao) [reply pinot-common/src/main/java/org/apache/pinot/common/metrics/BrokerGauge.java]: It represents how many table pinot broker is collecting the metrics for. The exact usage of the metrics will show up in later diffs

File: pinot-spi/src/main/java/org/apache/pinot/spi/data/Schema.java (hunk 4 above)
Comment (jamesyfshao) [reply pinot-spi/src/main/java/org/apache/pinot/spi/data/Schema.java]: right now we don't have support for multiple primary keys. Do you think it will be an important feature to add multiple primary key support right now or you think it is fine to postpone it to later diff?

File: pinot-spi/src/main/java/org/apache/pinot/spi/data/Schema.java (hunk 5 above)
Comment (jamesyfshao) [reply pinot-spi/src/main/java/org/apache/pinot/spi/data/Schema.java]: definitely would be better this way, will submit a change later to update this

File: pinot-spi/src/main/java/org/apache/pinot/spi/data/Schema.java (hunk 6 above)
Comment (jamesyfshao) [reply pinot-spi/src/main/java/org/apache/pinot/spi/data/Schema.java]: done.

It is used for converting a pinot column value into the appropriate binary representation for upsert primary key field. (make sure all primary key will become a byte array)

File: pom.xml (hunk 7 above)
Comment (jamesyfshao) [reply pom.xml]: moved to submodules

**Comment by jamesyfshao** [comment]
> Can you please link the design doc in the pr description so that it is easier to understand the concept of upsert?
> Since this pr is touching the core configs, we need to spend some time discussing how to design the config keys because they are very hard to change in the future

Linked the design doc in the description. also [google doc design doc](https://www.google.com/url?sa=t&rct=j&esrc=s&source=appssearch&uact=8&cd=0&cad=rja&q&sig2=4ypfJoFtqjSdua_7dziWCQ&ved=0ahUKEwjVhpqIybToAhXDHWIKHWFcBVs4ABABKAAwAA&url=https://drive.google.com/a/uber.com/open?id%3D1SFFir7ByxCff-aVYxQeTHpNhPXeP5q7P4g_6O2iNGgU%26usp%3Dchrome_omnibox&usg=AOvVaw3KEdQYFMPa7tgpOn01R5ol) with comments. We have discussed this a long time back but I definitely open to go through it again as things have already changed. Do you think we should do some remote meeting sync or something else?

File: pinot-common/src/main/java/org/apache/pinot/common/metrics/AbstractMetrics.java (hunk 10 above)
Comment (jamesyfshao) [reply pinot-common/src/main/java/org/apache/pinot/common/metrics/AbstractMetrics.java]: | Is this for real-time? We already have push type of either APPEND or REFRESH for offline; aggregateMetrics for real-time. What does UPSERT stand for?

replied in earlier comment, basically design for supporting update/insert in pinot realtime ingestion

File: pinot-common/src/main/java/org/apache/pinot/common/config/TableConfig.java (hunk 9 above)
Comment (jamesyfshao) [reply pinot-common/src/main/java/org/apache/pinot/common/config/TableConfig.java]: updated to follow the convention of rest of the code

File: pinot-common/src/main/java/org/apache/pinot/common/config/TableConfig.java (hunk 8 above)
Comment (jamesyfshao) [reply pinot-common/src/main/java/org/apache/pinot/common/config/TableConfig.java]: issue with merging, fixed. Thanks for pointing it out

File: pinot-spi/src/main/java/org/apache/pinot/spi/data/Schema.java (hunk 4 above)
Comment (kishoreg) [reply pinot-spi/src/main/java/org/apache/pinot/spi/data/Schema.java]: Better to design it for multiple, even if the implementation can handle only one primary key

**Comment by kishoreg** [comment]
One concern here would be that the update related configuration appears in multiple places. One thing that can simplify this PR is to add a UpdateConfig under TableConfig and move everything under that. That allows you to mark that config as beta/unstable and iterate quickly
- updatesemantic (or the new name)
- primarykey
- offsetkey
-

File: pinot-spi/src/main/java/org/apache/pinot/spi/data/Schema.java (hunk 4 above)
Comment (ChethanUK) [reply pinot-spi/src/main/java/org/apache/pinot/spi/data/Schema.java]: > don't have support for multiple primary keys

There will be some cases where multiple keys make sense right? Like in clickstream: `userId and timestamp`

**Comment by ChethanUK** [comment]
> also [google doc design doc](https://www.google.com/url?sa=t&rct=j&esrc=s&source=appssearch&uact=8&cd=0&cad=rja&q&sig2=4ypfJoFtqjSdua_7dziWCQ&ved=0ahUKEwjVhpqIybToAhXDHWIKHWFcBVs4ABABKAAwAA&url=https://drive.google.com/a/uber.com/open?id%3D1SFFir7ByxCff-aVYxQeTHpNhPXeP5q7P4g_6O2iNGgU%26usp%3Dchrome_omnibox&usg=AOvVaw3KEdQYFMPa7tgpOn01R5ol) with comments.

It's asking uber login. Can you change share settings?

**Comment by jamesyfshao** [comment]
> > also [google doc design doc](https://www.google.com/url?sa=t&rct=j&esrc=s&source=appssearch&uact=8&cd=0&cad=rja&q&sig2=4ypfJoFtqjSdua_7dziWCQ&ved=0ahUKEwjVhpqIybToAhXDHWIKHWFcBVs4ABABKAAwAA&url=https://drive.google.com/a/uber.com/open?id%3D1SFFir7ByxCff-aVYxQeTHpNhPXeP5q7P4g_6O2iNGgU%26usp%3Dchrome_omnibox&usg=AOvVaw3KEdQYFMPa7tgpOn01R5ol) with comments.
> 
> It's asking uber login. Can you change share settings?
> 
> > remote meeting sync or something else
> 
> @kishoreg At least monthly meeting and sync up on project status will be helpful to all...

Yeah our google doc account sharing setting is kind of wanky that it doesn't allow blanket share with everyone outside our organization. I hate it but it is really out of my control. However, just submit a request to view/comment and I can add you as soon as I see it.

We originally plan to have a meeting earlier this month but the pandemic kind of mess up with all of our plans. Maybe we can have a work-group specific remote meeting (eg, upsert meeting with kishore/jackie/chethan etc) in the meantime and do the pinot committee meeting later 

@Jackie-Jiang @kishoreg @ChethanUK let me know if you feel meeting is necessary or you prefer me to update the schema first and see if a remote meeting is necessary

File: pinot-common/src/main/java/org/apache/pinot/common/config/TableNameBuilder.java (hunk 2 above)
Comment (jamesyfshao) [reply pinot-common/src/main/java/org/apache/pinot/common/config/TableNameBuilder.java]: removed this method, @kishoreg this method is used in other components that I will submit PR later

File: pinot-common/src/main/java/org/apache/pinot/common/utils/CommonConstants.java [hunk 11]
```diff
@@ -308,6 +320,23 @@ public ServerType getServerType() {
     public static final String PREFIX_OF_CONFIG_OF_PINOT_CRYPTER = "crypter";
   }
 
+  public static class Grigio {
```
Comment (mcvsubbu) [inline pinot-common/src/main/java/org/apache/pinot/common/utils/CommonConstants.java:323]: Can we change the name of this class like I had suggested before?

File: pinot-common/src/main/java/org/apache/pinot/common/utils/CommonConstants.java [hunk 12]
```diff
@@ -378,4 +407,23 @@ public ServerType getServerType() {
     @Deprecated
     public static final String TABLE_NAME = "segment.table.name";
   }
+
+  public enum UpdateSemantic {
+    APPEND,
```
Comment (mcvsubbu) [inline pinot-common/src/main/java/org/apache/pinot/common/utils/CommonConstants.java:412]: Please add comments on each of these. What does APPEND mean, vs what UPSERT means. Feel free to point to the design doc

File: pinot-spi/src/main/java/org/apache/pinot/spi/data/Schema.java (hunk 4 above)
Comment (jamesyfshao) [reply pinot-spi/src/main/java/org/apache/pinot/spi/data/Schema.java]: @kishoreg @ChethanUK I have changed the primary key in schema config to be a list instead of a single string. For now we are still performing check to ensure that primary key will be one and only one. However we should be able to change that once we have mulitple-primary-key support in place

File: pinot-common/src/main/java/org/apache/pinot/common/config/TableConfig.java [hunk 13]
```diff
@@ -470,6 +505,18 @@ public void setInstanceAssignmentConfigMap(
...
 
+  public UpdateSemantic getUpdateSemantic() {
+    return _updateSemantic;
+  }
+
+  public void setUpdateSemantic(UpdateSemantic updateSemantic) {
+    _updateSemantic = updateSemantic;
+  }
+
+  public boolean isTableForUpsert() {
```
Comment (kishoreg) [inline pinot-common/src/main/java/org/apache/pinot/common/config/TableConfig.java:516]: let's avoid adding additional utility methods to config classes if possible

**Comment by jamesyfshao** [comment]
@kishoreg @mcvsubbu I have updated the diff per suggestion you guys gave in the last meeting, please feel free to take a look

@Jackie-Jiang I have updated the diff to make sure #1: it is only related table config changes required by pinot upsert feature (see [doc](https://docs.google.com/document/d/1SFFir7ByxCff-aVYxQeTHpNhPXeP5q7P4g_6O2iNGgU/edit?usp=sharing) for details) and #2 push other pinot-common related changes for pinot upsert features to later PRs

File: pinot-spi/src/main/java/org/apache/pinot/spi/config/IngestionModeConfig.java [hunk 14]
```diff
@@ -0,0 +1,152 @@
...
+  // primary key refers to the column name that is the primary key of this upsert table
+  private List<String> _primaryKeys;
+  // offset key refers to the column name that we are going to store the offset value to
+  private String _offsetKey;
+  // validFrom key refers to the column name that we are going to store the validFrom value to
+  private String _validFromKey;
+  // validFrom key refers to the column name that we are going to store the validFrom value to
+  private String _validUntilKey;
+
+  public static final String UPSERT_TABLE_CONFIG_VALUE = "upsert";
```
Comment (mcvsubbu) [inline pinot-spi/src/main/java/org/apache/pinot/spi/config/IngestionModeConfig.java:50]: we should not need these two for now

File: pinot-spi/src/main/java/org/apache/pinot/spi/config/IngestionModeConfig.java [hunk 15]
```diff
@@ -0,0 +1,152 @@
...
+    } else {
+      _primaryKeys = primaryKeys;
+    }
+    _offsetKey = offsetKey;
+    _validFromKey = validFromKey;
+    _validUntilKey = validUntilKey;
+
+    if (UPSERT_TABLE_CONFIG_VALUE.equals(_ingestionMode)) {
+      Preconditions.checkState(_primaryKeys.size() == 1,
+          "pinot upsert require one and only one primary key");
```
Comment (mcvsubbu) [inline pinot-spi/src/main/java/org/apache/pinot/spi/config/IngestionModeConfig.java:80]:
```suggestion
          "Upsert feature supports only one primary key");
```

File: pinot-spi/src/main/java/org/apache/pinot/spi/config/IngestionModeConfig.java [hunk 16]
```diff
@@ -0,0 +1,152 @@
...
+      Preconditions.checkState(StringUtils.isNotEmpty(_offsetKey),
+          "pinot upsert require one offset key");
```
Comment (mcvsubbu) [inline pinot-spi/src/main/java/org/apache/pinot/spi/config/IngestionModeConfig.java:82]:
```suggestion
          "Upsert feature requires \"offsetColumn\" to be set");
```

File: pinot-spi/src/main/java/org/apache/pinot/spi/config/IngestionModeConfig.java [hunk 17]
```diff
@@ -0,0 +1,152 @@
...
+      Preconditions.checkState(StringUtils.isNotEmpty(_validFromKey),
+          "pinot upsert require one validFrom key");
```
Comment (mcvsubbu) [inline pinot-spi/src/main/java/org/apache/pinot/spi/config/IngestionModeConfig.java:84]:
```suggestion
          "Upsert feature requires \"validFromColumn\" to be set");
```

File: pinot-spi/src/main/java/org/apache/pinot/spi/config/IngestionModeConfig.java [hunk 18]
```diff
@@ -0,0 +1,152 @@
...
+      Preconditions.checkState(StringUtils.isNotEmpty(_validUntilKey),
+          "pinot upsert require one validUntil key");
```
Comment (mcvsubbu) [inline pinot-spi/src/main/java/org/apache/pinot/spi/config/IngestionModeConfig.java:86]:
```suggestion
          "Upsert feature requires \"validUntilColumn\"  to be set");
```

File: pinot-spi/src/main/java/org/apache/pinot/spi/config/TableConfig.java [hunk 19]
```diff
@@ -44,6 +45,7 @@
   public static final String QUERY_CONFIG_KEY = "query";
   public static final String INSTANCE_ASSIGNMENT_CONFIG_MAP_KEY = "instanceAssignmentConfigMap";
   public static final String FIELD_CONFIG_LIST_KEY = "fieldConfigList";
+  public static final String INGESTION_MODE_CONFIG_KEY = "ingestionModeConfig";
```
Comment (mcvsubbu) [inline pinot-spi/src/main/java/org/apache/pinot/spi/config/TableConfig.java:48]:
```suggestion
  public static final String UPSERT_CONFIG_KEY = "upsertConfig";
```

File: pinot-spi/src/main/java/org/apache/pinot/spi/config/IngestionModeConfig.java [hunk 20]
```diff
@@ -0,0 +1,152 @@
...
+import com.google.common.collect.ImmutableList;
+import org.apache.commons.lang.StringUtils;
+import org.apache.pinot.spi.utils.JsonUtils;
+
+import javax.annotation.Nullable;
+import java.util.List;
+
+@SuppressWarnings("unused")
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class IngestionModeConfig {
```
[the remark this answers is not in the document]
Comment (Jackie-Jiang) [reply pinot-spi/src/main/java/org/apache/pinot/spi/config/IngestionModeConfig.java:37]: +1. I think this config is only for upsert feature, and not applicable for general ingestion. When this config exist, we can assume the upsert feature is on.

[the remark this answers is not in the document]
Comment (Jackie-Jiang) [reply pinot-spi/src/main/java/org/apache/pinot/spi/config/IngestionModeConfig.java:37]: Also, please extends BaseJsonConfig, which handles the Json ser-de and common methods like hashCode(), equals() and toString(). You can refer to `RoutingConfig` which is similar to this one.

File: pinot-spi/src/main/java/org/apache/pinot/spi/config/IngestionModeConfig.java [hunk 21]
```diff
@@ -0,0 +1,152 @@
...
+  // new config key to indicate if a table is for upsert. If this is define as upsert, it would be an upsert table
+  // for other value it would be append value for now
+  private String _ingestionMode;
+  // primary key refers to the column name that is the primary key of this upsert table
+  private List<String> _primaryKeys;
```
[the remark this answers is not in the document]
Comment (Jackie-Jiang) [reply pinot-spi/src/main/java/org/apache/pinot/spi/config/IngestionModeConfig.java:42]: +1. You can also make it final. Same for other places

File: pinot-spi/src/main/java/org/apache/pinot/spi/config/IngestionModeConfig.java [hunk 22]
```diff
@@ -0,0 +1,152 @@
...
+  public static final String APPEND_TABLE_CONFIG_VALUE = "append";
+
+  public static final IngestionModeConfig DEFAULT_APPEND_INGESTION_MODE =
+      new IngestionModeConfig(APPEND_TABLE_CONFIG_VALUE, null, null, null,
+          null);
+
+  @JsonCreator
+  public IngestionModeConfig(
+      @JsonProperty(value="ingestionMode") @Nullable String ingestionMode,
+      @JsonProperty(value="primaryKeys") @Nullable List<String> primaryKeys,
```
Comment (Jackie-Jiang) [inline pinot-spi/src/main/java/org/apache/pinot/spi/config/IngestionModeConfig.java:60]: If this config is for upsert only, I think all arguments are required

**Comment by jamesyfshao** [comment]
@mcvsubbu @Jackie-Jiang updated per your feedback and make the config to be used only if the table is for upsert and clean up on naming, please take a look when you have time

File: pinot-common/src/main/java/org/apache/pinot/common/utils/config/TableConfigUtils.java [hunk 23]
```diff
@@ -116,8 +117,16 @@ public static TableConfig fromZNRecord(ZNRecord znRecord)
       });
     }
 
+    UpsertConfig upsertConfig = null;
+    String upsertConfigString = simpleFields.get(TableConfig.UPSERT_CONFIG_KEY);
+    if (upsertConfigString != null) {
+      upsertConfig = JsonUtils.stringToObject(upsertConfigString, UpsertConfig.class);
+    }
+
```
Comment (Jackie-Jiang) [inline pinot-common/src/main/java/org/apache/pinot/common/utils/config/TableConfigUtils.java:125]: (nit) extra empty line

File: pinot-common/src/test/java/org/apache/pinot/common/utils/config/TableConfigSerDeTest.java [hunk 24]
```diff
@@ -231,6 +234,17 @@ public void testSerDe()
       assertEquals(tableConfigToCompare, tableConfig);
       checkFieldConfig(tableConfigToCompare);
     }
+    {
+      // with upsert config
+      UpsertConfig upsertConfig = new UpsertConfig(ImmutableList.of("pk"), "offset",
```
Comment (Jackie-Jiang) [inline pinot-common/src/test/java/org/apache/pinot/common/utils/config/TableConfigSerDeTest.java:239]: Prefer using `Collections.singletonList()`

File: pinot-common/src/test/java/org/apache/pinot/common/utils/config/TableConfigSerDeTest.java [hunk 25]
```diff
@@ -367,4 +381,12 @@ private void checkFieldConfig(TableConfig tableConfig) {
     assertNull(secondFieldConfig.getIndexType());
     assertNull(secondFieldConfig.getProperties());
   }
+
+  private void checkTableConfigWithUpsertConfig(TableConfig tableConfig, TableConfig tableConfigToCompare) {
```
Comment (Jackie-Jiang) [inline pinot-common/src/test/java/org/apache/pinot/common/utils/config/TableConfigSerDeTest.java:385]: Can you make this similar to other config check? Pass in a  table config and check the actual value inside the config. Then serialize and deserialize (both string and ZNRecord) and check again

File: pinot-spi/src/test/java/org/apache/pinot/spi/config/UpsertConfigTest.java [hunk 26]
```diff
@@ -0,0 +1,72 @@
...
+      fail();
+    } catch (RuntimeException ex) {}
+
+    try {
+      upsertConfig = new UpsertConfig(ImmutableList.of("pk1"),
+          "offset", "validFrom", null);
+      fail();
+    } catch (RuntimeException ex) {}
+  }
+}
```
Comment (Jackie-Jiang) [inline pinot-spi/src/test/java/org/apache/pinot/spi/config/UpsertConfigTest.java:72]: (nit) empty line

**Comment by jamesyfshao** [comment]
hi @mcvsubbu we have updated the pr following the comments. Appreciate if you can take a look to see if anything else you feel we need to update further

File: pinot-spi/src/main/java/org/apache/pinot/spi/config/UpsertConfig.java [hunk 27]
```diff
@@ -0,0 +1,82 @@
...
+      @JsonProperty(value="validFromColumn") String validFromColumn,
+      @JsonProperty(value="validUntilColumn") String validUntilColumn) {
+
+    _primaryKeyColumns = primaryKeyColumns;
+    _offsetColumn = offsetColumn;
+    _validFromColumn = validFromColumn;
+    _validUntilColumn = validUntilColumn;
+
+    Preconditions.checkState(_primaryKeyColumns.size() == 1,
+        "upsert feature support only one primary key column");
```
Comment (mcvsubbu) [inline pinot-spi/src/main/java/org/apache/pinot/spi/config/UpsertConfig.java:58]:
```suggestion
        "Upsert feature supports only one primary key column");
```

File: pinot-spi/src/main/java/org/apache/pinot/spi/config/UpsertConfig.java [hunk 28]
```diff
@@ -0,0 +1,82 @@
...
+    Preconditions.checkState(StringUtils.isNotEmpty(_offsetColumn),
+        "upsert feature requires \"offsetColumn\" to be set ");
```
Comment (mcvsubbu) [inline pinot-spi/src/main/java/org/apache/pinot/spi/config/UpsertConfig.java:60]:
```suggestion
        "Upsert feature requires \"offsetColumn\" to be set ");
```

File: pinot-spi/src/main/java/org/apache/pinot/spi/config/UpsertConfig.java [hunk 29]
```diff
@@ -0,0 +1,82 @@
...
+    Preconditions.checkState(StringUtils.isNotEmpty(_validFromColumn),
+        "upsert feature requires \"validFromColumn\" to be set");
```
Comment (mcvsubbu) [inline pinot-spi/src/main/java/org/apache/pinot/spi/config/UpsertConfig.java:62]:
```suggestion
        "Upsert feature requires \"validFromColumn\" to be set");
```

File: pinot-spi/src/main/java/org/apache/pinot/spi/config/UpsertConfig.java [hunk 30]
```diff
@@ -0,0 +1,82 @@
...
+    Preconditions.checkState(StringUtils.isNotEmpty(_validUntilColumn),
+        "upsert feature requires \"validUntilColumn\" to be set");
```
Comment (mcvsubbu) [inline pinot-spi/src/main/java/org/apache/pinot/spi/config/UpsertConfig.java:64]:
```suggestion
        "Upsert feature requires \"validUntilColumn\" to be set");
```

**Comment by jamesyfshao** [comment]
> minor comments, other than that, lgtm thanks

merged the comments diff, thanks for checking on those details. Do you think any further changes you want? It would be great if you can change your review status from requesting-changes so I can start merging it in

**Comment by mcvsubbu** [comment]
@jamesyfshao I am good with this PR, thanks.
